### PR TITLE
First implementation of `build_evt()` and `build_skm()`

### PR DESCRIPTION
**Pull request by patgo25**
event tier builder to build a static event level from dsp, hit, and tcm files with help of json configuration file.

### CODE REVIEW

**Comment by gipert** [comment]
My suggestions about the config field names, to be further discussed (there is certainly still room for improvement):

```js
{
  "channels_defs": {
    "geds_on": ["ch1084803", "ch1084804", "ch1121600"], // explicit
    "my_spms": "legendmeta.get_on_spms_channels('20220629T095300Z')", // function provided by a 'legendmeta' subpackage
    "pulser": "legendmeta.get_pulser_channel('20220629T095300Z')" // same
  },
  "operations": {
    "multiplicity": {  // a scalar
      "channels": "geds_on",
      "hit_aggregation_mode": "sum", // sum booleans (see expression)
      "store_channel_id": true,
      "expression": "hit.cuspEmax_ctc_cal > a", // a boolean
      "parameters": {"a": 25},
      "default": 0
    },
    "energy": { // a variable-length array
      "channels": "geds_on",
      "aggregation_mode": "collect", // or just "none"/omitted?
      "query": "hit.cuspEmax_ctc_cal > 25", // or "where" instead of "query" (numpy)?
      "store_channel_id": true,
      "expression": "hit.cuspEmax_ctc_cal" // the hit field itself
    },
    "tot_energy": { // a scalar
      "channels": "geds_on",
      "aggregation_mode": "collect",
      "query": "hit.cuspEmax_ctc_cal > 25"
      "expression": "hit.cuspEmax_ctc_cal",
    },
    "aoe": { // a variable-length array
      "aggregation_mode": "keep_at:energy_id", // keep same channel ids as in field "energy"
      "expression": "hit.AoE_Classifier"
    },
    "is_pulser": {
      "channels": "pulser",
      "aggregation_mode": "logical_or", // omittable since only one channel?
      "expression": "dsp.trapTmax > 200"
    },
    "is_global_ringing": {
      "channels": "geds_on",
      "aggregation_mode": "logical_or", // OR
      "expression": "~(is_valid_neg_current)",
      "default": false
    },
    "is_global_physical": {
      "channels": "geds_on",
      "aggregation_mode": "logical_and", // AND
      "expression": "(is_valid_0vbb) & (is_valid_tail) & (is_valid_baseline) & ~(is_noise_burst) & ~(is_negative)",
      "default": false
    },
    "lar_multiplicity": {
      "channels": "spms_on",
      "mode": "function",
      "expression": "pygama.evt.modules.spm.get_majority(0.5, evt.t0, 48000, 1000, 5000)"
    },
    "is_lar_rejected": {
      "expression": "(evt.lar_energy >4) | (evt.lar_multiplicity > 4)" // use other already-computed evt outputs
    },
}
```

### Further features for discussion

- in the `build_hit()` config we reference parameters in the expression with `@`. We should be consistent: we can decide to drop the `@`, but the `build_hit()` config syntax should be updated accordingly
- do we want to have an `outputs` array at the top of the config, to make clear which fields get written to disk (see `build_dsp()` and `build_hit()` configs)?
- Should we promote definition of channel rawid columns to its own config block? In the end these are columns like the others and we might want to set custom column names (more details in next comment)

**Comment by patgo25** [comment]
An config example:
```
{
                "channels": {
                    "geds_on": ["ch1084803", "ch1084804", "ch1121600"],
                     "spms_on": {
                          "module": "pygama.evt.modules.legend_meta.legend_meta",
                          "system": "spms",
                          "usability": "on",
                          "meta_path": "/data2/public/prodenv/prod-blind/ref/v02.00/inputs"
                      },
                    "muon": "ch1027202",
                },
                "operations": {
                    "energy":{
                        "channels": "geds_on",
                        "aggregation_mode": "vectorize",
                        "query": "hit.cuspEmax_ctc_cal>25",
                        "expression": "hit.cuspEmax_ctc_cal"
                    },
                    "energy_id":{
                        "channels": "geds_on",
                        "aggregation_mode": "vectorize",
                        "query": "hit.cuspEmax_ctc_cal>25",
                        "expression": "tcm.array_id"
                    },
                    "is_muon_rejected":{
                        "channels": "muon",
                        "aggregation_mode": "any",
                        "expression": "dsp.wf_max>a",
                        "parameters": {"a":15100},
                        "initial": false
                    },
                    "multiplicity":{
                        "channels":  ["geds_on","geds_no_psd","geds_ac"],
                        "aggregation_mode": "sum",
                        "expression": "hit.cuspEmax_ctc_cal > a",
                        "parameters": {"a":25},
                        "initial": 0
                    },
                    "t0":{
                        "aggregation_mode": "keep_at:evt.energy_id",
                        "expression": "dsp.tp_0_est"
                    },
                    "lar_energy":{
                        "channels": "spms_on",
                        "aggregation_mode": "function",
                        "expression": ".modules.spm.get_energy(0.5,evt.t0,48000,1000,5000)"
                    },
                }
            }
```

**Comment by patgo25** [comment]
TODO frome dev-call:
implement VoV sorting

**Comment by gipert** [comment]
@patgo25 you should now be able to use all latest pydataobj features here, if you merge `main` into this PR.

File: src/pygama/evt/build_evt.py [hunk 1]
```diff
@@ -0,0 +1,1567 @@
...
+    f_hit: str,
+    f_dsp: str,
+    chns: list,
+    chns_rm: list,
+    expr: str,
+    exprl: list,
+    qry: str | np.ndarray,
+    nrows: int,
+    var_ph: dict = None,
+    defv=np.nan,
```
Comment (gipert) [inline src/pygama/evt/build_evt.py:835]: Type annotation?

File: src/pygama/evt/build_evt.py [hunk 2]
```diff
@@ -0,0 +1,1567 @@
...
+       dictionary of evt and additional parameters and their values
+    defv
+       default value
+    """
+
+    out = np.full(len(ch_comp.nda), defv, dtype=type(defv))
+
+    for ch in np.unique(ch_comp.nda.astype(int)):
+        # skip default value
+        if f"ch{ch}" not in store.ls(f_hit):
```
Comment (gipert) [inline src/pygama/evt/build_evt.py:943]:
```suggestion
        if f"ch{ch}" not in lh5.ls(f_hit):
```

File: src/pygama/evt/build_evt.py [hunk 3]
```diff
@@ -0,0 +1,1567 @@
...
+    idx: np.ndarray,
+    ids: np.ndarray,
+    f_hit: str,
+    f_dsp: str,
+    expr: str,
+    exprl: list,
+    ch_comp: VectorOfVectors,
+    chns_rm: list,
+    var_ph: dict = None,
+    defv=np.nan,
```
Comment (gipert) [inline src/pygama/evt/build_evt.py:976]: Type annotation?

File: src/pygama/evt/build_evt.py [hunk 4]
```diff
@@ -0,0 +1,1567 @@
...
+    chns_rm
+       list of channels to be skipped from evaluation and set to default value
+    var_ph
+       dictionary of evt and additional parameters and their values
+    defv
+       default value
+    """
+
+    # blow up vov to aoesa
+    out = ch_comp.to_aoesa().nda
```
Comment (gipert) [inline src/pygama/evt/build_evt.py:1006]:
```suggestion
    out = ch_comp.to_aoesa().view_as("np")
```

Comment (gipert) [reply src/pygama/evt/build_evt.py:1006]: Same here: I guess here one could do it with Awkward (`.view_as("ak")`)?

File: src/pygama/evt/build_evt.py [hunk 5]
```diff
@@ -0,0 +1,1567 @@
...
+
+    wo_mode
+        writing mode
+    group
+        lh5 root group name
+    tcm_group
+        lh5 root group in tcm file
+    """
+
+    lstore = store.LH5Store()
```
Comment (gipert) [inline src/pygama/evt/build_evt.py:1308]:
```suggestion
    lstore = LH5Store()
```

File: src/pygama/evt/build_evt.py [hunk 6]
```diff
@@ -0,0 +1,1567 @@
...
+            chns[k] = met(v | attr)
+
+        elif isinstance(v, str):
+            chns[k] = [v]
+
+        elif isinstance(v, list):
+            chns[k] = [e for e in v]
+
+    nrows = len(
+        store.load_nda(f_tcm, ["cumulative_length"], tcm_group)["cumulative_length"]
```
Comment (gipert) [inline src/pygama/evt/build_evt.py:1358]: Assuming you have `store = LH5Store()` somewhere above:
```suggestion
        store.read(f_tcm, ["cumulative_length"], tcm_group).view_as("np")
```

File: src/pygama/evt/build_evt.py [hunk 7]
```diff
@@ -0,0 +1,1567 @@
...
+    evt_group
+        lh5 root group of the evt file
+    """
+
+    if wo_mode not in ["o", "overwrite", "n", "new"]:
+        raise ValueError(
+            wo_mode
+            + " is a invalid writing mode. Valid options are: 'o', 'overwrite','n','new'"
+        )
+    lstore = store.LH5Store()
```
Comment (gipert) [inline src/pygama/evt/build_evt.py:1510]:
```suggestion
    lstore = LH5Store()
```

File: src/pygama/evt/build_evt.py [hunk 8]
```diff
@@ -0,0 +1,1567 @@
...
+    fields = store.ls(f_evt, evt_group)
```
Comment (gipert) [inline src/pygama/evt/build_evt.py:1511]:
```suggestion
    fields = lh5.ls(f_evt, evt_group)
```

File: src/pygama/evt/build_evt.py [hunk 9]
```diff
@@ -0,0 +1,1567 @@
...
+        for e in store.ls(f_evt, evt_group)
+        if e.split("/")[-1] in exprl
+    ]
+    var = {e: lstore.read(evt_group + e, f_evt)[0] for e in flds}
+
+    # to make any operations to VoVs we have to blow it up to a table (future change to more intelligant way)
+    arr_keys = []
+    for key, value in var.items():
+        if isinstance(value, VectorOfVectors):
+            var[key] = value.to_aoesa().nda
```
Comment (gipert) [inline src/pygama/evt/build_evt.py:1528]:
```suggestion
            var[key] = value.to_aoesa().view_as("np")
```

File: src/pygama/evt/build_evt.py [hunk 10]
```diff
@@ -0,0 +1,1567 @@
...
+        elif isinstance(value, Array):
+            var[key] = value.nda
```
Comment (gipert) [inline src/pygama/evt/build_evt.py:1530]:
```suggestion
            var[key] = value.view_as("np")
```

**Review by gipert — COMMENTED** [review]
I left some comments about pydataobj related renamings, but did not have time to go through all the code...

File: src/pygama/evt/modules/legend.py [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+"""
+from importlib import import_module
+
+
+def metadata(params: dict) -> list:
+    # only import legend meta data when needed.
+    # LEGEND collaborators can use the meta keyword
+    # While for users w/o access to the LEGEND meta data this is still working
+    lm = import_module("legendmeta")
+    lmeta = lm.LegendMetadata(path=params["meta_path"])
```
Comment (gipert) [inline src/pygama/evt/modules/legend.py:12]: Please allow for environment variable expansion in the config, you can just pass the expression to:

https://github.com/legend-exp/legend-pydataobj/blob/main/src/lgdo/lh5/utils.py

We should not hardcode paths here: https://github.com/legend-exp/legend-dataflow-config/pull/33

File: src/pygama/evt/modules/spm.py [hunk 12]
```diff
@@ -0,0 +1,352 @@
...
+    lim,
+    trgr,
+    tdefault,
+    tmin,
+    tmax,
+    swin,
+    trail,
+    min_first_pls_ene,
+) -> Array:
+    # ignore stupid numpy warnings
```
Comment (gipert) [inline src/pygama/evt/modules/spm.py:250]: Didn't you say you fixed this?

File: src/pygama/skm/build_skm.py [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+
+def build_skm(
+    f_evt: str,
+    f_hit: str,
+    f_dsp: str,
+    f_tcm: str,
+    f_skm: str,
+    skm_conf: dict | str,
+    wo_mode="w",
+    skim_format: str = "parquet",
```
Comment (gipert) [inline src/pygama/skm/build_skm.py:28]: The default should be LH5 here. I also think that this should be the only supported option, users can use `view_as()` to get the data in some other format.

File: src/pygama/skm/build_skm.py [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+        - ``overwrite`` or ``o``: replaces existing file.
+
+    skim_format
+        data format of the skimmed output (``hdf``, ``lh5`` or ``parquet``).
+    group
+        LH5 root group name (only used if ``skim_format`` is ``lh5``).
+    """
+    f_dict = {"evt": f_evt, "hit": f_hit, "dsp": f_dsp, "tcm": f_tcm}
+    log = logging.getLogger(__name__)
+    log.debug(f"I am skimning {len(f_evt) if isinstance(f_evt,list) else 1} files")
```
Comment (gipert) [inline src/pygama/skm/build_skm.py:103]:
```suggestion
    log.debug(f"I am skimming {len(f_evt) if isinstance(f_evt,list) else 1} files")
```

**Review by gipert — CHANGES_REQUESTED** [review]
I see harcoded tier names or reliance on specific file naming conventions, can we make it more agnostic? What happens if we add another tier or change name (e.g. `pht` tier)? I like more the idea of being more explicit about things in the config file.

File: src/pygama/evt/modules/legend.py (hunk 11 above)
Comment (patgo25) [reply src/pygama/evt/modules/legend.py]: Done in  b83c362

File: src/pygama/skm/build_skm.py (hunk 13 above)
Comment (patgo25) [reply src/pygama/skm/build_skm.py]: Resolved in 542c53b

File: src/pygama/skm/build_skm.py (hunk 14 above)
Comment (patgo25) [reply src/pygama/skm/build_skm.py]: Resolved in 542c53b

File: src/pygama/evt/modules/spm.py (hunk 12 above)
Comment (patgo25) [reply src/pygama/evt/modules/spm.py]: Cleaned up in 542c53b

Comment (patgo25) [reply src/pygama/evt/modules/spm.py]: and b02be96

**Comment by patgo25** [comment]
pre-commit.ci autofix

**Review by gipert — APPROVED** [review]

**Comment by gipert** [comment]
God forgive me for what I have done

**Pull request merged**